R/backend-common.R

```diff
@@ -94,131 +88,53 @@ transform_to_tensor.WaveMC <- function(
+  audio <- as(audio, "Wave")
   transform_to_tensor(audio, out, normalization, channels_first)
 }
 
 #' @export
 transform_to_tensor.av <- function(
-  audio,
-  out = NULL,
-  normalization = TRUE,
-  channels_first = TRUE
-) {
-
+    audio,
+    out = NULL,
+    normalization = TRUE,
+    channels_first = TRUE) {
   sample_rate <- attr(audio, "sample_rate")
   out_tensor <- torch::torch_tensor(audio, dtype = torch::torch_float())
 
-  if(!channels_first)
-    out_tensor = out_tensor$t()
-
-  # normalize if needed
-  if(is.null(normalization)) normalization <- TRUE
-  internal__normalize_audio(out_tensor, normalization)
-
-  return(list(out_tensor, sample_rate))
-}
-
-to_tensor_float <- function(x) torch::torch_tensor(x, dtype = torch::torch_float())
-
-#' @export
-transform_to_tensor.audiofile <- function(
-  audio,
-  out = NULL,
-  normalization = FALSE,
-  channels_first = TRUE
-) {
-
-  sample_rate <- audio$sample_rate
-
-  if(audio$channels > 1) {
-    out_tensor <- Map(to_tensor_float, audio$waveform)
-    out_tensor <- torch::torch_stack(out_tensor)
-  } else {
-    out_tensor <- to_tensor_float(audio$waveform[[1]])$unsqueeze(1)
+  if (!channels_first) {
+    out_tensor <- out_tensor$t()
   }
 
-  if(!channels_first)
-    out_tensor = out_tensor$t()
-
   # normalize if needed
-  if(is.null(normalization)) normalization <- FALSE
-  if(is.logical(normalization) && isTRUE(normalization)) {
-    bits <- audio$bit %||% 32
-    normalization <- 2^(bits-1)
-  }
+  if (is.null(normalization)) normalization <- TRUE
   internal__normalize_audio(out_tensor, normalization)
 
   return(list(out_tensor, sample_rate))
 }
 
-#' MP3 Information
-#'
-#' Retrive metadata from mp3 without load the audio samples in memory.
-#'
-#' @param filepath (chr) path to mp3 file
-#'
-#' @return AudioMetaData: sample_rate, channels, samples
-#'
-#' @examples
-#' mp3_path <- system.file("sample_audio_1.mp3", package = "torchaudio")
-#' mp3_info(mp3_path)
-#'
-#' @export
-mp3_info <- function(filepath) {
-  info <- get_info_mp3(filepath)
-  AudioMetaData$new(
-    sample_rate = info$hz,
-    num_frames = info$samples,
-    num_channels = info$channels
-  )
-}
-
-#' Wave Information
-#'
-#' Retrive metadata from wav without load the audio samples in memory.
+#' Audio Information
 #'
-#' @param filepath (chr) path to wav file
+#' Retrieve audio metadata.
 #'
-#' @return AudioMetaData: sample_rate, channels, samples
+#' @param filepath (str) path to the audio file.
+#' @return AudioMetaData: an R6 class with fields sample_rate, channels, samples.
 #'
 #' @examples
-#' wav_path <- system.file("waves_yesno/1_1_0_1_1_0_1_1.wav", package = "torchaudio")
-#' wav_info(wav_path)
+#' path <- system.file("waves_yesno/1_1_0_1_1_0_1_1.wav", package = "torchaudio")
+#' torchaudio_info(path)
 #'
 #' @export
-wav_info <- function(filepath) {
-  info <- tuneR::readWave(filepath, header = TRUE)
+torchaudio_info <- function(filepath) {
+  audio <- av::read_audio_bin(filepath)
```

Comment:
Does this read the entire audio just to query some information? Is there a way to just read the headers? Since this is called by the `av_loader` basically all files are being read twice which sounds like something we don't want to do in datasets code.

Reply:
@dfalbel  I re-checked the code and really, there's  a better way!
The reason that the updated `torchaudio_info()` uses `read_audio_bin()` is that it returns the number of samples, which is not part of what `av_media_info()` reports (but it's useful I think).
Then in `av_loader(),` I overlooked that I just need the sample rate, which is part of the above! So here, I can just use `av_media_info() ` :-)

This also means that relative overhead is now incurred only when you just want the info, not when reading files/datasets, which is awesome. 

Thanks for getting me to look at this again!

Reply:
Awesome! Looks great to me!

Reply:
Thank you @dfalbel, merging!